azurelinuxagent/ga/agent_update.py

```diff
@@ -0,0 +1,577 @@
+import glob
+import json
+import os
+import shutil
+import time
+
+from azurelinuxagent.common import conf, logger
+from azurelinuxagent.common.event import add_event, WALAEventOperation
+from azurelinuxagent.common.exception import AgentUpgradeExitException, UpdateError
+from azurelinuxagent.common.future import ustr
+from azurelinuxagent.common.logger import LogLevel
+from azurelinuxagent.common.protocol.extensions_goal_state import GoalStateSource
+from azurelinuxagent.common.protocol.restapi import VMAgentUpdateStatuses, VMAgentUpdateStatus
+from azurelinuxagent.common.utils import fileutil, textutil
+from azurelinuxagent.common.utils.flexible_version import FlexibleVersion
+from azurelinuxagent.common.version import CURRENT_VERSION, AGENT_NAME, AGENT_DIR_PATTERN
+from azurelinuxagent.ga.exthandlers import HandlerManifest
+
+AGENT_ERROR_FILE = "error.json"  # File name for agent error record
+AGENT_MANIFEST_FILE = "HandlerManifest.json"
+MAX_FAILURE = 3  # Max failure allowed for agent before declare bad agent
+
+
+def get_agent_update_handler(protocol):
+    return AgentUpdateHandler(protocol)
+
+
+class PersistentUpdateHandlerState(object):
+    """
+    This class is primarily used to maintain the in-memory persistent state for the agent updates.
+    This state will be persisted throughout the current service run and might be modified by external classes.
+    """
+    report_error = None
+
+
+class AgentUpdateHandler(object):
+
+    def __init__(self, protocol):
+        self.__protocol = protocol
+        self.__ga_family = conf.get_autoupdate_gafamily()
+        self.__autoupdate_enabled = conf.get_autoupdate_enabled()
+        self.__gs_id = self.protocol.get_goal_state().incarnation
+        self._last_attempted_update_time = None
+        self._last_attempted_update_version = FlexibleVersion("0.0.0.0")
+        self._pkg_list = None
+
+    @property
+    def protocol(self):
+        return self.__protocol
+
+    @property
+    def ga_family(self):
+        return self.__ga_family
+
+    @property
+    def autoupdate_enabled(self):
+        return self.__autoupdate_enabled
+
+    @property
+    def gs_id(self):
+        return self.__gs_id
+
+    def __is_update_allowed(self, requested_version):
+        """
+        check to see if update is allowed once per (as specified in the conf.get_autoupdate_frequency())
+        """
+        now = time.time()
+
+        if self._last_attempted_update_time is not None and self._last_attempted_update_version == requested_version:
```

Comment:
I don't quite get this check. If we attempt the update and fail then self._last_attempted_update_version == requested_version would be True and we would keep incrementing the next time, possibly never retrying the update, wouldn't we?

Reply:
we don't increment the actual last_attenped_update_time and compute next_time always from last_attenped_update_time.

If we don't reach the autoupdate frequency, we don't update the last_attenped_update_time and version variables.

Reply:
Say 

1) we get a goal state with a new requested version. We update self._last_attempted_update_version to that version, but later the update fails. 

2) we get another goal state with the same requested version. self._last_attempted_update_version == requested_version is True so we increment next_attempt_time. If next_attempt_time > now we return False

then 2) keeps happening over and over and we never update

Reply:
If the auto update frequency set to 1hr. Then we don't update till that 1hr(that's the idea right). Once time elapsed will try attempt update. I don't know what we are missing here.

Reply:
i see it now, your are ok